### PR TITLE
Support Calendar Access Levels (iOS17+)

### LINKED ISSUE #1108: [Bug]: In iOS 17 , can't request Full Access , this is the latest Calendar access levels in iOS 17.
### Please check the following before submitting a new issue.

- [X] I have searched the [existing issues](https://github.com/baseflow/flutter-permission-handler/issues).
- [X] I have carefully [read the documentation](https://github.com/Baseflow/flutter-permission-handler/blob/main/permission_handler/README.md) and verified I have added the required platform specific configuration.

### Please select affected platform(s)

- [ ] Android
- [X] iOS
- [ ] Windows

### Steps to reproduce

1. In iOS 17 : https://developer.apple.com/documentation/technotes/tn3152-migrating-to-the-latest-calendar-access-levels 
2. I can't request Full Access

### Expected results

Full Access can be granted by requesting.

### Actual results

Only can request Add Events Only

### Code sample

<details><summary>Code sample</summary>

```dart

var status = await Permission.calendar.request();

```

</details>

### Screenshots or video

<details>
<summary>Screenshots or video demonstration</summary>

[Upload media here]
image

</details>

### Version

  permission_handler: ^10.4.3

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.10.1, on macOS 13.4.1 22F82 darwin-x64, locale zh-Hans-CN)
    • Flutter version 3.10.1 on channel stable at /Users/xuxu/flutter/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision d3d8effc68 (8 weeks ago), 2023-05-16 17:59:05 -0700
    • Engine revision b4fb11214d
    • Dart version 3.0.1
    • DevTools version 2.23.1

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.1)
    • Android SDK at /Users/xuxu/Library/Android/sdk
    • Platform android-33, build-tools 33.0.1
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.15+0-b2043.56-8887301)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E222b
    • CocoaPods version 1.12.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.15+0-b2043.56-8887301)

[✓] IntelliJ IDEA Ultimate Edition (version 2022.2.3)
    • IntelliJ at /Users/xuxu/Applications/JetBrains Toolbox/IntelliJ IDEA Ultimate.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] IntelliJ IDEA Ultimate Edition (version 2022.2.3)
    • IntelliJ at /Users/xuxu/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/222.4345.14/IntelliJ IDEA.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] Connected device (3 available)
    • iPhone 14 Pro Max (mobile) • 89E998A2-CAB6-4A9B-B4B1-AB81DD5D4347 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-17-0 (simulator)
    • macOS (desktop)            • macos                                • darwin-x64     • macOS 13.4.1 22F82 darwin-x64
    • Chrome (web)               • chrome                               • web-javascript • Google Chrome 114.0.5735.198

[✓] Network resources
    • All expected network resources are available.

• No issues found!

```

</details>

### PR DESCRIPTION
**Pull request by MilosKarakas**
Fixes #1108.

I don't know if you agree with my approach, I can change it if you have any suggestions.

## Pre-launch Checklist

- [x] I made sure the project builds.
- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is does not need version changes.
- [x] I updated `CHANGELOG.md` to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I rebased onto `main`.
- [x] I added new tests to check the change I am making, or this PR does not need tests.
- [x] I made sure all existing and new tests are passing.
- [x] I ran `dart format .` and committed any changes.
- [x] I ran `flutter analyze` and fixed any errors.

[Contributor Guide]: https://github.com/Baseflow/flutter-permission-handler/blob/master/CONTRIBUTING.md
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

### CODE REVIEW

File: permission_handler_platform_interface/lib/src/permissions.dart
Comment (JeroenWeener) [inline permission_handler_platform_interface/lib/src/permissions.dart:39]: We prefer adding new permissions at the end of the list and giving them a higher id, instead of shuffling ids around. While your approach is nicer to the maintainers eye, it is usually a source of bugs because we forget to sync the ids between dart and the platform code. Also, we'd need to update all the platform packages too.

File: permission_handler_platform_interface/pubspec.yaml
Comment (JeroenWeener) [inline permission_handler_platform_interface/pubspec.yaml:6]:
```suggestion
version: 3.12.0
```

File: permission_handler_apple/pubspec.yaml
Comment (JeroenWeener) [inline permission_handler_apple/pubspec.yaml:5]:
```suggestion
version: 9.2.0
```

File: permission_handler/CHANGELOG.md
Comment (JeroenWeener) [inline permission_handler/CHANGELOG.md:1]: Typically we have CHANGELOG entries for all different packages (`permission_handler`, `permission_handler_apple`, `permission_handler_platform_interface`).

**Review by JeroenWeener — CHANGES_REQUESTED** [review]
Thank you for your contribution! We are very happy to have you onboard and are impressed by your PR 😄

I have some remarks regarding the architecture. I would appreciate it if you could resolve these issues. If you do not have time/motivation please let us know so we can take over and make sure your works makes it in.

Instead of an `add only` permission, we would prefer a `full access` permission. This would change the permission approach such that `add only` is the default (by requesting `Permission.calendar`) and that for full access, you'd need `Permission.calendarFullAccess`.

Because there will be two separate permissions for calendar that users of the plugins can use, we do not need extra permission statuses such as 'write only'. Typically, we want to avoid adding more statuses, as this corrodes the experience of our users. In this case, a user that gets 'granted' on `Permission.calendar` knows that they have write-only access, so there is no need for additional statuses.

This PR requires changes to all platforms, so the new permission is implemented everywhere. In this case, the new permission on platforms other than iOS can just use the same logic as the current `Permission.calendar`, as there is no distinction on those platforms between full access and write-only access. There should be updates to the different platform packages (`permission_handler_android`, `permission_handler_windows`, `permission_handler_web` and `permission_handler_linux`). If it is more convenient, we can also make these changes.

**Comment by MilosKarakas** [comment]
@JeroenWeener Thank you for kind words, I've been using your package for a long time so it made sense to give something back. 

All of requested changes make sense. I have time this week to readjust the PR, but if I run into any issues while adjusting the packages for platforms other then iOS I'll reach out for help.

**Comment by MilosKarakas** [comment]
@JeroenWeener Does it make sense to mark this in the changelog as a breaking change since requesting calendar now degrades the permission to a write-only, while calendarFullAccess works as calendar used to?

**Comment by JeroenWeener** [comment]
@MilosKarakas That makes sense 👍

**Comment by dawsonbristow20** [comment]
Need this update soon! Our app is pending app store release due to this issue😓

**Comment by MilosKarakas** [comment]
@dawsonbristow20 I had a lot of work during last week so I didn't manage to complete this PR, I'm working on this today.

**Comment by MilosKarakas** [comment]
@JeroenWeener Hi, I think I've adjusted everything as requested, you can review it again when you find time. I wasn't sure only about this part, can you give me an example of what you expect here:

> There should be updates to the different platform packages (permission_handler_android, permission_handler_windows, permission_handler_web and permission_handler_linux). If it is more convenient, we can also make these changes.

**Comment by parker-sherrill** [comment]
@MilosKarakas Thank you for this!

**Comment by MilosKarakas** [comment]
@parker-sherrill Hey, no problem.. Just make sure to wait until it's reviewed and officially released on pub.dev before using it, I can't guarantee everything is 100% right before @JeroenWeener  or someone other who's authorised confirms it.

**Comment by parker-sherrill** [comment]
Any update on review? @JeroenWeener

**Comment by JeroenWeener** [comment]
Apologies for the radio silence. We had an internal discussion about splitting read-only and full access calendar permissions. Since Android natively also supports this split, we want to deprecate the `calendar` permission and introduce two new permissions: `calendarReadOnly` and `calendarFullAccess`. I will put up a PR for the platform interface and the Android package later today or tomorrow. Then, this PR can be updated, adhering to the new platform implementation.

**Comment by JeroenWeener** [comment]
Supporting issues: #1181, #1182.

**Comment by JeroenWeener** [comment]
The platform interface has been updated (see #1181).

@MilosKarakas could you update your dependency to `permission_handler_platform_interface` to versions `3.12.0`?

There are now three permissions to take into account; `calendar`, `calendarReadOnly` and `calendarFullAccess`. The `calendar` permission will be deprecated in the future. For now, it will function identical to `calendarFullAccess`.

This means that the changes in this PR are no longer breaking. If everything went well, you should be able to dismiss your changes to the platform interface. Regarding my earlier comments about updating other platforms, don't worry about it. We will take care of Android and the other platforms.

If you have any questions, feel free to hit me up.

**Comment by MilosKarakas** [comment]
@JeroenWeener Hi, I'll be done with changes by the end of the week, as soon as I catch some time from work. 

I'll let you know if any questions arise.

File: permission_handler_apple/pubspec.yaml
Comment (JeroenWeener) [inline permission_handler_apple/pubspec.yaml:21]:
```suggestion
  permission_handler_platform_interface: ^4.0.0
```

File: permission_handler_apple/CHANGELOG.md
Comment (JeroenWeener) [inline permission_handler_apple/CHANGELOG.md:2]:
```suggestion
* Adds the support for `Permission.calendarWriteOnly` and `Permission.calendarFullAccess` permissions which are introduced in iOS 17+.
```

File: permission_handler/CHANGELOG.md
Comment (JeroenWeener) [inline permission_handler/CHANGELOG.md:5]:
```suggestion
* Updates `permission_handler_platform_interface` dependency to version 4.0.0.
```

File: permission_handler/CHANGELOG.md
Comment (JeroenWeener) [inline permission_handler/CHANGELOG.md:3]:
```suggestion
* on iOS 17 use `Permission.calendarWriteOnly` to request a write-only access to the calendar. For full access to calendar use `Permission.calendarFullAccess`. 
```

**Review by JeroenWeener — CHANGES_REQUESTED** [review]
My apologies: I discovered this week by reading through your PR, that the iOS permissions are split in `WRITE` and `FULL_ACCESS`, as opposed to `READ` and `FULL_ACCESS`. I updated the platform interface to use this new split (version 4.0.0). Could you update the PR to accommodate for this? This split should make more sense, as it is in line with the way permissions are split on iOS.

**Comment by JeroenWeener** [comment]
Hi @MilosKarakas!

Do you still have time to work on this PR in the near future?
I'd be happy to offer help where needed.

**Comment by MilosKarakas** [comment]
Hi @JeroenWeener 

Thanks for the updates. I'm absent until next week when I would be able to adjust the PR and hopefully finalise it. If that timeline seems fine with you that we can proceed that way.

@tp I'm glad it works fine, mention of that change in the changelog makes sense to me. Thanks for pointing it out. Regarding the error you suggested that does sound good but I think that would require changes to a lot more places in the codebase, which in my opinion is out of scope of this PR.

**Comment by MilosKarakas** [comment]
@JeroenWeener I applied the changes, let me know if any additional changes are required for this PR to get merged.

File: permission_handler_apple/ios/Classes/strategies/EventPermissionStrategy.m
Comment (JeroenWeener) [inline permission_handler_apple/ios/Classes/strategies/EventPermissionStrategy.m:72]: Forgive me as my Objective-C knowledge could use some polishing, but would this work if the application wants the option to ask for either permission? Let's say `PERMISSION_EVENTS_FULL_ACCESS` and `PERMISSION_EVENTS` are both `1`, and the application requests `Permission.calendarWriteOnly`.

**Review by JeroenWeener — CHANGES_REQUESTED** [review]
Thanks @MilosKarakas!

I left you some small comments. @TimHoogstrate could you have a look at the code as well? Since you have more experience with Objective-C.

File: permission_handler/CHANGELOG.md
Comment (TimHoogstrate) [inline permission_handler/CHANGELOG.md:2]:
```suggestion
* Adds support for iOS 17+ [Calendar access levels](https://developer.apple.com/documentation/technotes/tn3152-migrating-to-the-latest-calendar-access-levels)
```

File: permission_handler/CHANGELOG.md
Comment (TimHoogstrate) [inline permission_handler/CHANGELOG.md:3]:
```suggestion
* On iOS 17+ use `Permission.calendarWriteOnly` to request a write-only access to the calendar. For full access to calendar use `Permission.calendarFullAccess`. 
```

File: permission_handler_apple/ios/Classes/strategies/EventPermissionStrategy.m
Comment (TimHoogstrate) [inline permission_handler_apple/ios/Classes/strategies/EventPermissionStrategy.m:126]: Is it always either one? Because otherwise you could be missing an "else" return.

**Comment by mvanbeusekom** [comment]
Solves #1202

File: permission_handler_apple/ios/Classes/strategies/EventPermissionStrategy.m
Comment (MilosKarakas) [reply permission_handler_apple/ios/Classes/strategies/EventPermissionStrategy.m]: I went with assumption that it's always either one, which was wrong, it might be both at the same time. I think the latest commit addresses this issue.

**Comment by MilosKarakas** [comment]
@JeroenWeener  @TimHoogstrate I've adjusted EventPermissionStrategy.m to address your comments. I tested it and it works fine on my demo app. If after review you find any issues you can take over since I've dragged it for long now, and wouldn't like to stall this PR further. Thanks for the reviews.

**Comment by JeroenWeener** [comment]
Thanks @MilosKarakas.

Appreciate the effort you have put into this. I will take a look and make sure it gets merged 👍

**Comment by TimHoogstrate** [comment]
One note: I've tested the implementation and when I run the permission_handler_apple Example app on iOS 17.1. And I request calendarWriteOnly. I am unable to request calendarFullAccess. I think it should show something else. Let's see how we can fix this before merging. When I click it i get an error: 

```[VERBOSE-2:dart_vm_initializer.cc(41)] Unhandled Exception: PlatformException(ERROR_ALREADY_REQUESTING_PERMISSIONS, A request for permissions is already running, please wait for it to finish before doing another request (note that you can request multiple permissions at the same time)., null, null)
#0      StandardMethodCodec.decodeEnvelope (package:flutter/src/services/message_codecs.dart:652:7)
#1      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:310:18)
<asynchronous suspension>
#2      MethodChannelPermissionHandler.requestPermissions (package:permission_handler_platform_interface/src/method_channel/method_channel_permission_handler.dart:79:9)
<asynchronous suspension>
#3      _PermissionState.requestPermission (package:permission_handler_apple_example/main.dart:144:20)
<asynchronous suspension>```

**Comment by TimHoogstrate** [comment]
It seems to work correctly, The apple example project was not updated by adding the full access macro's. After I added the Full access macro's it was working as a charm.

**Comment by MilosKarakas** [comment]
@JeroenWeener  Thanks, I hope it gets merged.

@TimHoogstrate Thanks for taking a look, I'm glad it works fine.

**Comment by JeroenWeener** [comment]
Update: last week we merged updates to the different platforms. I just updated this PR which will expose these platform changes.

@TimHoogstrate could you provide another review?

**Review by TimHoogstrate — APPROVED** [review]
LGTM but i added some comments to make it more in line with other parts of the documentation

**Review by JeroenWeener — APPROVED** [review]

**Comment by JeroenWeener** [comment]
@MilosKarakas we merged your PR just now. The new calendar permissions for iOS should be up and running starting with `permission_handler` version `11.1.0`. Once again thank you for contributing!

**Pull request merged**
